### ISSUE TITLE
Faulty 304 handling results in `Failed to construct 'Response'` error

### ISSUE THREAD

**Issue by halfbakedsneed** [kettanaito/msw#1893]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 14 or higher

### Browsers

- Chrome (119.0.6045.123)
- Firefox (120.0)

### Reproduction repository

https://github.com/halfbakedsneed/msw-bad-304-handling

### Reproduction steps

```
pnpm install
pnpm start
```

Browse to http://localhost:3000 and hit "Click Me"

### Current behavior

For "bypassed" requests, responses with a `304` status code will result in the following being thrown:

```
Failed to construct 'Response': Response with null body status cannot have body
TypeError: Failed to construct 'Response': Response with null body status cannot have body
    at http://localhost:3000/static/js/bundle.js:48332:69
    at ServiceWorkerContainer.<anonymous> (http://localhost:3000/static/js/bundle.js:48595:15)
```

### Expected behavior

The service worker script correctly treats responses with a `204` status code [differently](https://github.com/mswjs/msw/blob/main/src/mockServiceWorker.js#L124-L128) when handling responses. The same behaviour should be exhibited when handling `304` responses, such that an error is not thrown.

See https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304

>The response must not contain a body...

**Comment by mattcosta7** [comment]
@halfbakedsneed I tried your reproduction, and it appears that ts-node isn't properly running your api, so the passthrough is failing because it's getting a 404 I think?

```
[1] TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /Users/matthewcostabile/msw-bad-304-handling/src/api.ts
[1]     at new NodeError (node:internal/errors:406:5)
[1]     at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:99:9)
[1]     at defaultGetFormat (node:internal/modules/esm/get_format:142:36)
[1]     at defaultLoad (node:internal/modules/esm/load:120:20)
[1]     at nextLoad (node:internal/modules/esm/hooks:833:28)
[1]     at load (/Users/matthewcostabile/msw-bad-304-handling/node_modules/.pnpm/ts-node@10.9.1_@types+node@16.18.65_typescript@4.9.5/node_modules/ts-node/dist/child/child-loader.js:19:122)
[1]     at nextLoad (node:internal/modules/esm/hooks:833:28)
[1]     at Hooks.load (node:internal/modules/esm/hooks:416:26)
[1]     at MessagePort.handleMessage (node:internal/modules/esm/worker:168:24)
[1]     at [nodejs.internal.kHybridDispatch] (node:internal/event_target:807:20) {
[1]   code: 'ERR_UNKNOWN_FILE_EXTENSION'
[1] }
```

I believe that is likely giving an empty 404 page response body, which is likely not what's expected?

**Comment by AriPerkkio** [comment]
Could this issue be re-opened until the linked PRs are finished?

I'm seeing this same error when using MSW with Cypress. I can reproduce this issue 100% of times locally. Below is screenshot of MSW's execution stopped in debugger right before the error is thrown. You can see the contents of `n` on the right side: 

![image](https://github.com/mswjs/msw/assets/14806298/c9548a41-10dc-4aab-b0a2-883865346784)

I can only see this error coming up when using Cypress.

**Comment by kettanaito** [comment]
## Released: v2.0.10 🎉

This has been released in v2.0.10!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.0.10)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.0.10)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Comment by AriPerkkio** [comment]
Fix verified with `msw@2.0.10`. Thanks @mattcosta7! 🙌

**Comment by diegolaciar** [comment]
Using MSW v2.3.1 

Doing like this on my resolver :

```javascript
return HttpResponse.json({}, { status: 204 });
```

I get an error 500 from MSW:

```javascript
api-usuarios-validaciones-nuevos-usuarios.response.resolver.ts:12 [MSW] Uncaught exception in the request handler for "GET http://bff-obj.k8sds.gscorp.ad/v1.0/usuarios/validaciones-nuevos-usuarios?cuil=2":

TypeError: Failed to construct 'Response': Response with null body status cannot have body
    at new HttpResponse (http://localhost:4200/vendor.js:191286:5)
    at HttpResponse.json (http://localhost:4200/vendor.js:191320:12)
    at http://localhost:4200/main.js:6125:60
    at Generator.next (<anonymous>)
    at fulfilled (http://localhost:4200/vendor.js:100948:56)
    at _ZoneDelegate.invoke (http://localhost:4200/polyfills.js:7680:158)
    at Zone.run (http://localhost:4200/polyfills.js:7463:35)
    at http://localhost:4200/polyfills.js:8556:28
    at _ZoneDelegate.invokeTask (http://localhost:4200/polyfills.js:7707:171)
    at Zone.runTask (http://localhost:4200/polyfills.js:7509:37)```
